flow/connectors/connelasticsearch/elasticsearch.go

```diff
@@ -78,101 +88,178 @@ func (esc *ElasticsearchConnector) Close() error {
+					}
+				}
 			}
 		}
 	}()
 
-	for qRecord := range stream.Records {
-		qRecordJsonMap := make(map[string]any)
+	var docId string
+	var bulkIndexFatalError error
+	var bulkIndexErrors []error
+	var bulkIndexOnFailureMutex sync.Mutex
 
-		if upsertColIndex >= 0 {
-			docId = fmt.Sprintf("%v", qRecord[upsertColIndex].Value())
-		} else {
-			docId = uuid.New().String()
+	for record := range req.Records.GetRecords() {
+		var bodyBytes []byte
+		var err error
+		action := actionIndex
+
+		switch record.(type) {
+		case *model.InsertRecord[model.RecordItems], *model.UpdateRecord[model.RecordItems]:
+			bodyBytes, err = recordItemsProcessor(record.GetItems())
+			if err != nil {
+				esc.logger.Error("[es] failed to json.Marshal record", slog.Any("error", err))
+				return nil, fmt.Errorf("[es] failed to json.Marshal record: %w", err)
+			}
+		case *model.DeleteRecord[model.RecordItems]:
+			action = actionDelete
+			// no need to supply the document since we are deleting
+			bodyBytes = nil
 		}
-		for i, field := range schema.Fields {
-			switch r := qRecord[i].(type) {
-			// JSON is stored as a string, fix that
-			case qvalue.QValueJSON:
-				qRecordJsonMap[field.Name] = json.RawMessage(shared.
-					UnsafeFastStringToReadOnlyBytes(r.Val))
-			default:
-				qRecordJsonMap[field.Name] = r.Value()
+
+		bulkIndexer, ok := esBulkIndexerCache[record.GetDestinationTableName()]
+		if !ok {
+			bulkIndexer, err = esutil.NewBulkIndexer(esutil.BulkIndexerConfig{
+				Index:  record.GetDestinationTableName(),
+				Client: esc.client,
+				// can't really ascertain how many tables present to provide a reasonable value
+				NumWorkers:    1,
+				FlushInterval: 10 * time.Second,
+			})
+			if err != nil {
+				esc.logger.Error("[es] failed to initialize bulk indexer", slog.Any("error", err))
+				return nil, fmt.Errorf("[es] failed to initialize bulk indexer: %w", err)
 			}
+			esBulkIndexerCache[record.GetDestinationTableName()] = bulkIndexer
 		}
-		qRecordJsonBytes, err := json.Marshal(qRecordJsonMap)
-		if err != nil {
-			esc.logger.Error("[es] failed to json.Marshal record", slog.Any("error", err))
-			return 0, fmt.Errorf("[es] failed to json.Marshal record: %w", err)
+
+		if len(req.TableNameSchemaMapping[record.GetDestinationTableName()].PrimaryKeyColumns) == 1 {
+			qValue, err := record.GetItems().GetValueByColName(
+				req.TableNameSchemaMapping[record.GetDestinationTableName()].PrimaryKeyColumns[0])
+			if err != nil {
+				esc.logger.Error("[es] failed to process record", slog.Any("error", err))
+				return nil, fmt.Errorf("[es] failed to process record: %w", err)
+			}
+			docId = fmt.Sprint(qValue.Value())
+		} else {
+			tablePkey, err := model.RecToTablePKey(req.TableNameSchemaMapping, record)
+			if err != nil {
+				esc.logger.Error("[es] failed to process record", slog.Any("error", err))
+				return nil, fmt.Errorf("[es] failed to process record: %w", err)
+			}
+			docId = hex.EncodeToString(tablePkey.PkeyColVal[:])
```

Comment:
why not base64?

Reply:
there is a limit to size of the document ID, so won't work for really wide pkeys

Reply:
right, but I'm suggesting using `base64.NoPadding.EncodeToString` instead of `hex.EncodeToString`, as base64 is a more compact encoding (6 bits per character) compared to hex (4 bits per character)

Reply:
makes sense, changed